get_block_templates() must return unique templates/template parts

anton-vlasenko opened the pull request:
<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

This PR aims to fix a bug in the `get_block_templates()` method.
`array_column` doesn't work correctly with an array of objects on PHP <= 5.6. So, under certain circumstances, `get_block_templates()` could return entities with the same ID (slug).
A regression unit test (included in this PR) should prevent this from happening in the future.

Trac ticket: https://core.trac.wordpress.org/ticket/56271
Steps to test this PR: https://core.trac.wordpress.org/ticket/56271#comment:18

Props to @darrencoutts118 for the original fix (see https://github.com/WordPress/wordpress-develop/pull/3010).
Props to @costdev for code reviews/valuable feedback.
My contribution was to write unit tests to make sure that the bug would not happen again.

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

anton-vlasenko committed:
Implement a regression unit test to ensure that get_block_templates() returns unique values on PHP 5.6 and below.

anton-vlasenko committed:
Add a test method for testing template parts.

anton-vlasenko committed:
Refactor Tests_Blocks_GetBlockTemplates to use the data provider.

anton-vlasenko committed:
Fix bug in Tests_Blocks_GetBlockTemplates::test_it_returns_unique_entities.

anton-vlasenko committed:
Fix bug in Tests_Blocks_GetBlockTemplates::data_test_it_returns_unique_entities.

anton-vlasenko committed:
Fix bug in Tests_Blocks_GetBlockTemplates::test_it_returns_unique_entities.

anton-vlasenko committed:
Refactoring: rename some variables.

anton-vlasenko committed:
Use wp_list_pluck() instead of the anonymous function.

anton-vlasenko committed:
Refactor Tests_Blocks_GetBlockTemplates::test_it_returns_unique_entities.

anton-vlasenko committed:
Fix typo.

anton-vlasenko committed:
Add a new assertion to ensure the test works correctly if the original template gets deleted or removed for some reason.

anton-vlasenko committed:
Fix typo.

anton-vlasenko committed:
Use wp_list_pluck() instead of array_column().
array_column doesn't correctly work with objects on PHP 5.6 and below.

costdev commented on tests/phpunit/tests/blocks/getBlockTemplates.php:
```diff
@@ -0,0 +1,107 @@
+	 *
+	 * @return array
+	 */
+	public function data_test_it_returns_unique_entities() {
```
Data providers should appear below the test methods that use them.

anton-vlasenko replied:
Fixed in 950b02d853e104593d9f733672e3c21f9d065f64

costdev commented on tests/phpunit/tests/blocks/getBlockTemplates.php:
```diff
@@ -0,0 +1,107 @@
+	 *
+	 * @param string $template_type The template type.
+	 * @param string $original_template_id ID (slug) of the default entity.
+	 * @param string $error_message An error message to display if the test fails.
```
```suggestion
	 * @param string $template_type        The template type.
	 * @param string $original_template_id ID (slug) of the default entity.
	 * @param string $error_message        An error message to display if the test fails.
```

The alignment of the description strings is a little off.

anton-vlasenko replied:
`composer run format` doesn't seem to fix these indents for some reason.
Fixed in b15a8a4449a98a1cf4dcffa5869c00c3b2de3276

costdev commented on tests/phpunit/tests/blocks/getBlockTemplates.php:
```diff
@@ -0,0 +1,107 @@
+
+	public static function wpSetUpBeforeClass() {
+
+		// This template has to have the same ID ("block-theme/index") as the template that is shipped with the
+		// "block-theme" theme. This is needed for testing purposes.
```
```
/*
 * This template has to have the same ID ("block-theme/index") as the template that is shipped with the
 * "block-theme" theme. This is needed for testing purposes.
 */
```

anton-vlasenko replied:
Fixed in b15a8a4449a98a1cf4dcffa5869c00c3b2de3276

costdev commented on tests/phpunit/tests/blocks/getBlockTemplates.php:
```diff
@@ -0,0 +1,107 @@
+		);
+		wp_set_post_terms( static::$template->ID, static::TEST_THEME, 'wp_theme' );
+
+		// This template part has to have the same ID ("block-theme/small-header") as the template part that is shipped with the
+		// "block-theme" theme. This is needed for testing purposes.
```
```
/*
 * This template part has to have the same ID ("block-theme/small-header") as the template part that is shipped with the
 * "block-theme" theme. This is needed for testing purposes.
 */
```

anton-vlasenko replied:
Fixed in b15a8a4449a98a1cf4dcffa5869c00c3b2de3276

costdev reviewed (changes requested):
Thanks @anton-vlasenko! This review is targeted at formatting and organisation of methods.

anton-vlasenko committed:
Fix code style (code review feedback).

anton-vlasenko committed:
Move the data provider below.
See https://github.com/WordPress/wordpress-develop/pull/3051#discussion_r936023683

anton-vlasenko committed:
Fix code style.

anton-vlasenko commented:
> Thanks @anton-vlasenko! This review is targeted at formatting and organisation of methods.

Thanks for the review, @costdev!
I've replied to your comments.

costdev reviewed (approved):
LGTM :+1:

ironprogrammer commented on tests/phpunit/tests/blocks/getBlockTemplates.php:
```diff
@@ -0,0 +1,112 @@
+<?php
+/**
+ * @group blocks
```
As the existing [tests/block-template-utils.php](https://github.com/WordPress/wordpress-develop/blob/eafd5201cb2aa295531830745e0960a12c415af8/tests/phpunit/tests/block-template-utils.php) is part of `@group block-templates`, should this test be part of that group?

I'm not sure where/how the "rules" regarding groups are determined, so I honestly don't know.

🙌🏻 Kudos for putting this into a separate test file!

anton-vlasenko replied:
Thanks for the review, @ironprogrammer.
Good catch!
Fixed in 66946f6df5fe9b950503aa38df30c416f6ed2f30

ironprogrammer commented on tests/phpunit/tests/blocks/getBlockTemplates.php:
```diff
@@ -0,0 +1,112 @@
+	 * @param string $original_template_id ID (slug) of the default entity.
+	 * @param string $error_message        An error message to display if the test fails.
+	 */
+	public function test_it_returns_unique_entities( $template_type, $original_template_id, $error_message ) {
```
Rather than referring to "entities", this method name might be clearer by referring to "templates", the subject of the test. E.g.:
`test_it_returns_unique_entities()`
`data_test_it_returns_unique_entities()`
could change to:
`test_return_unique_templates()`
`data_test_return_unique_templates()`

anton-vlasenko replied:
This test method tests not only templates (`post_type === wp_template`), but also template parts (`post_type === wp_template_part`). For this reason, I would leave the current name.

ironprogrammer commented on tests/phpunit/tests/blocks/getBlockTemplates.php:
```diff
@@ -0,0 +1,112 @@
+	 */
+	private static $template_part;
+
+	public static function wpSetUpBeforeClass() {
```
[Per the Core Handbook](https://make.wordpress.org/core/handbook/testing/automated-testing/writing-phpunit-tests/#shared-fixtures), should these set up/tear down methods be updated to their `snake_case` alternates for better compatibility?

anton-vlasenko replied:
Fixed in 66946f6df5fe9b950503aa38df30c416f6ed2f30

SergeyBiryukov replied:
Just to clarify, that handbook paragraph specifically applies to the native PHPUnit methods that need to be replaced with  WordPress' `snake_case` counterparts, as the tests would not pass across all supported PHP and PHPUnit versions otherwise:
* `setUpBeforeClass()` → `set_up_before_class()`
* `tearDownAfterClass()` → `tear_down_after_class()`


It does not apply to WordPress' own methods (note the `wp` prefix):
* `wpSetUpBeforeClass()`
* `wpTearDownAfterClass()`

These are fine to use and are used in many other tests, much more often than the previous methods. I would actually like to standardize on these across the test suite for consistency, since they are already used in the majority of tests.

Related: [#55918](https://core.trac.wordpress.org/ticket/55918)

ironprogrammer commented on tests/phpunit/tests/blocks/getBlockTemplates.php:
```diff
@@ -0,0 +1,112 @@
+		wp_set_post_terms( self::$template_part->ID, static::TEST_THEME, 'wp_theme' );
+	}
+
+	public static function wpTearDownAfterClass() {
```
See above comment.

anton-vlasenko replied:
Fixed in 66946f6df5fe9b950503aa38df30c416f6ed2f30

anton-vlasenko committed:
Refactoring:
1. Fix the test group. See https://github.com/WordPress/wordpress-develop/pull/3051/files#r941495218
2. Use WordPress methods instead of PHPUnit methods. See https://github.com/WordPress/wordpress-develop/pull/3051/files#r941909345 and https://github.com/WordPress/wordpress-develop/pull/3051/files#r941909448

SergeyBiryukov commented:
Thanks for the PR! Merged in [r53927](https://core.trac.wordpress.org/changeset/53927), backported to the 6.0 branch in [r53928](https://core.trac.wordpress.org/changeset/53928).
